### PR TITLE
Bootstrap 4 Upgrade

### PR DESCRIPTION
**Pull request by MaathavanJkr**

### CODE REVIEW

**Comment by sushain97** [comment]
Some fishy things:

image

These aren't aligning correctly.

image

Margins are broken here.

image

This is opening in the wrong direction.

Haven't tested anything but translation yet.

**Comment by MaathavanJkr** [comment]
> These aren't aligning correctly.

I dont get what isnt aligned correctly

**Comment by sushain97** [comment]
The drop-down and the other three buttons aren't correctly aligned. Try comparing it against how those buttons look on the current website.

**Comment by MaathavanJkr** [comment]
> The drop-down and the other three buttons aren't correctly aligned. Try comparing it against how those buttons look on the current website.

Ohh!, I didnt see it. :D . I think i need some sharp eyes.

**Comment by sushain97** [comment]
Errr, now the rounded corners are messed up.

before:

image

after:

image

Also, do you think you could host this somewhere with all the modes available so I could check it out and test the others look good? Copying the config from turkic.apertium.org probably works. Execute `config` in the JS console to see it.

File: assets/css/general.css [hunk 1]
```diff
@@ -38,43 +48,69 @@ h2 {
...
+/* Utilities */
+
+/* sass-lint:disable class-name-format */
+.w-0 {
+    width: 0%;
 }
 
-.font-weight-bold {
-    font-weight: bold;
+.border-0 {
```
Comment (sushain97) [inline assets/css/general.css]: I think Bootstrap 4 includes this?

File: assets/css/general.css [hunk 2]
```diff
@@ -223,12 +268,41 @@ html[dir='rtl'] .clearButton {
...
+
+html[dir='rtl'] .btn-group > .btn:last-child:not(.dropdown-toggle), /* sass-lint:disable-line class-name-format */
+.btn-group > .btn-group:last-child > .btn { /* sass-lint:disable-line class-name-format */
+    border-radius: .2rem 0 0 .2rem;
+}
+
+
+html[dir='rtl'] .btn-group > .btn:first-child:not(.dropdown-toggle), /* sass-lint:disable-line class-name-format */
+.btn-group > .btn-group:first-child > .btn { /* sass-lint:disable-line class-name-format */
+    border-radius: 0 .2rem .2rem 0;
```
Comment (sushain97) [inline assets/css/general.css]: I'm guessing that your change that fixed the alignment resulted in these styles no longer applying, breaking the round corners.

**Comment by MaathavanJkr** [comment]
> Also, do you think you could host this somewhere with all the modes available so I could check it out and test the others look good? Copying the config from turkic.apertium.org probably works. Execute `config` in the JS console to see it.

yeah, Sure, :)

**Comment by MaathavanJkr** [comment]
Is round corner messed for u on both ltr and rtl? its messed only in rtl for me thats why.

**Comment by MaathavanJkr** [comment]
And the rtl pages when reloading results a messed up alignmnent and ui. its because of the bootstrap 4.x rtl version css, its working properly in bootstrap rtl 3.x version.

File: assets/css/general.css (hunk 1 above)
Comment (MaathavanJkr) [reply assets/css/general.css]: Yup.

File: assets/css/general.css (hunk 2 above)
Comment (MaathavanJkr) [reply assets/css/general.css]: Yeahh. Fixed.

**Comment by sushain97** [comment]
> And the rtl pages when reloading results a messed up alignmnent and ui. its because of the bootstrap 4.x rtl version css, its working properly in bootstrap rtl 3.x version.

Oh, yep...

image

**Comment by MaathavanJkr** [comment]
You can use [https://maathavanjkr.github.io/](url) to check. i will be updating it. :)

**Comment by MaathavanJkr** [comment]
> Oh, yep...

What can we do? rtl i think i have to change each css for rtl or else we can use Bootstrap 3.x. So what to do?

**Comment by sushain97** [comment]
I would try figuring out why it works initially and then breaks on reload.

**Comment by MaathavanJkr** [comment]
The Page Breaks when the page is loaded. It works initially because the change of Locale when we change that the page isn't fully changed I think. I think that the RTL bootstrap CSS file is loaded only when the RTL locale is loaded without redirrecting. I'm not sure about this I have to check more clearly

**Comment by sushain97** [comment]
>  I think that the RTL bootstrap CSS file is loaded only when the RTL locale is loaded without redirrecting. I'm not sure about this I have to check more clearly

Cool. Looking forward to hearing back.

File: html/index.html.in [hunk 3]
```diff
@@ -73,15 +75,15 @@
...
             if ($('#faChecker').css('fontFamily') !== 'FontAwesome') {
                 $('<link type="text/css" rel="stylesheet" href="./css/font-awesome.min.css">').appendTo('head');
                 console.warn('FontAwesome CSS from CDN not available - reverting to local copy');
             }
         </script>
 
-        <span class="pull-right flip hide" id="rtlChecker"></span>
+        <span class="float-right flip" id="rtlChecker" style="display: none;"></span>
         <script type="text/javascript">
             if ($('#rtlChecker').css('float') !== 'left' && !$('#rtlStylesheet').attr('disabled')) {
```
Comment (MaathavanJkr) [inline html/index.html.in]: @sushain97 This thing causes the error. When its loaded initially `!$('#rtlStylesheet').attr('disabled')` this statement isnt satisfied as `localizeinteface()` js function, which changes the `#rtlStylesheet` attr doesnot happen when changing from one locale to another locale. When Its reloaded the LocalizeInteface function runs and the statement statisfies and the rtl css is loaded.

**Comment by MaathavanJkr** [comment]
i dont think we need a rtl bootstrap css as we already set features for rtl alone. I checked it without bootstrap rtl and its working properly.

**Comment by sushain97** [comment]
Hmmm... yeah. It does seem fine 🤔 I'm sure we initially added it for some reason but maybe that reason is no longer valid?

@jonorthwash can you see if you can find anything broken at https://maathavanjkr.github.io/test using RTL locales?

**Comment by jonorthwash** [comment]
@sushain97, RTL looks okay to me.  Of course, things don't translate or analyse, but I guess that's not a bootstrap issue?

**Comment by sushain97** [comment]
Hmm, it shouldn't be... I'll clone locally and try it out against turkic before merging.

**Comment by sushain97** [comment]
Everything seems to work! I'll merge this, delete any IE < 9 shims that we maintain and release as v4.0.0 since Bootstrap 4 no longer supports IE < 9.

**Comment by sushain97** [comment]
Ah, darn, I had to revert this on master after remembering to check mobile. There seem to be some issues here, particularly with the dropdown:

image

**Comment by sushain97** [comment]
(for whatever reason, Chrome desktop's mobile emulation doesn't render those language dropdowns correctly but my phone does so I'm not worried about that)

**Comment by MaathavanJkr** [comment]
#356

**Comment by MaathavanJkr** [comment]
> @sushain97, RTL looks okay to me. Of course, things don't translate or analyse, but I guess that's not a bootstrap issue?

Thats not an issue i think. its my fault when building.

**Comment by wei2912** [comment]
Some of the links overlap on horizontal phone/vertical tablet layouts:
2019-12-26-232919_scrot

**Comment by MaathavanJkr** [comment]
> Some of the links overlap on horizontal phone/vertical tablet layouts (I tried 812x375, iPhone X/XS on Firefox 71):
> 
> Screen Shot 2019-12-26 at 23 31 11

[](https://github.com/apertium/apertium-html-tools/pull/356/commits/89f524188d5394e017836fd0f0850aa83ebb74e8)

**Pull request merged**